tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs

```diff
@@ -170,15 +170,18 @@ public void InjectPathwayContextAsBase64String<TCarrier>(PathwayContext? context
         try
         {
             var previousContext = parentPathway;
-            if (previousContext == null && PreviousCheckpoint.Value != null && PreviousCheckpoint.Value.Kind != checkpointKind)
+            if (previousContext == null && LastConsumePathway.Value != null && checkpointKind == CheckpointKind.Produce)
             {
-                previousContext = PreviousCheckpoint.Value.Context;
+                // We only enter here on produce: when we consume, the only thing that matters is the parent we'd have read from the inbound message, not what happened before.
+                // We want to use the context from the previous consume (but we'll give priority to the parent passed in param if set).
+                previousContext = LastConsumePathway.Value;
```

Comment:
I think both cases are covered here:
 - in the case of "manual checkpointing" (something that is not offered currently because the checkpointing methods are `internal`), it is possible to pass the parent manually as well as parameter
  - when reading from upstream, the context would be stored in `LastConsumePathway` because reading = consume checkpoint, so this code works as intended